java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/refactoring/MoveRefactoring.java

```diff
@@ -116,101 +135,354 @@
+                    Exceptions.printStackTrace(
+                        Exceptions.attachSeverity(ex, Level.SEVERE)
+                    );
+                } else {
+                    client.showMessage(new MessageParams(MessageType.Error, ex.getLocalizedMessage()));
                 }
-            } else {
-                throw new IllegalArgumentException(String.format("Illegal number of arguments received for command: %s", command));
             }
-        } catch (Exception ex) {
-            client.showMessage(new MessageParams(MessageType.Error, ex.getLocalizedMessage()));
         }
-        return CompletableFuture.completedFuture(true);
-    }
 
-    private void move(NbCodeLanguageClient client, Object source, QuickPickItem target, ClasspathInfo info) {
-        try {
-            org.netbeans.modules.refactoring.api.MoveRefactoring refactoring;
-            if (source instanceof String) {
-                FileObject file = Utils.fromUri((String) source);
-                refactoring = new org.netbeans.modules.refactoring.api.MoveRefactoring(Lookups.fixed(file));
-                refactoring.getContext().add(JavaRefactoringUtils.getClasspathInfoFor(file));
-            } else {
-                TreePathHandle tph = (TreePathHandle) source;
-                refactoring = new org.netbeans.modules.refactoring.api.MoveRefactoring(Lookups.fixed(tph));
-                refactoring.getContext().add(tph.getKind() == Tree.Kind.CLASS ? JavaRefactoringUtils.getClasspathInfoFor(tph.getFileObject()) : new JavaMoveMembersProperties(tph));
+        private static Project getSelectedProject(NamedPath selectedProject) {
+            try {
+                String path = selectedProject.getPath();
+                return path != null ? FileOwnerQuery.getOwner(Utils.fromUri(path)) : null;
+            } catch (MalformedURLException ex) {
+                return null;
             }
-            if (target.getDescription() != null) {
-                refactoring.setTarget(Lookups.singleton(new URL((String) target.getUserData())));
-            } else {
-                ElementHandle handle = gson.fromJson(gson.toJson(target.getUserData()), ElementData.class).toHandle();
-                refactoring.setTarget(Lookups.singleton(TreePathHandle.from(handle, info)));
+        }
+
+        private static FileObject getSelectedRoot(NamedPath selectedRoot) {
+            try {
+                String path = selectedRoot.getPath();
+                return path != null ? Utils.fromUri(path) : null;
+            } catch (MalformedURLException ex) {
+                return null;
             }
-            client.applyEdit(new ApplyWorkspaceEditParams(perform(refactoring, "Move")));
-        } catch (Exception ex) {
-            client.showMessage(new MessageParams(MessageType.Error, ex.getLocalizedMessage()));
+        }
+    }
+
+    @Model(className = "ElementUI", instance = true, properties = {
+        @Property(name = "selected", type = boolean.class),
```

Comment:
Safer in the presence of [NETBEANS-6478](https://issues.apache.org/jira/browse/NETBEANS-6478).